File: src/main/java/cn/afeibaili/mchat/socket/Heartbeat.java
package cn.afeibaili.mchat.socket;

import java.io.Closeable;
import java.util.function.Consumer;

/**
 * 心跳
 *
 * @author AfeiBaili
 * @version 2026/3/3 01:16
 */

public class Heartbeat implements Closeable {
    Runnable action;
    Consumer<Exception> cat;

    Thread heartbeatThread;

    public Heartbeat(Runnable action, Consumer<Exception> cat) {
        this.action = action;
        this.cat = cat;

        heartbeatThread = new Thread(() -> {
            try {
                while (!Thread.currentThread().isInterrupted()) {
                    action.run();
                    Thread.sleep(60000 * 3);
                }
            } catch (Exception e) {
                cat.accept(e);
            }
        });
        heartbeatThread.start();
    }

    @Override
    public void close() {
        heartbeatThread.interrupt();
    }
}

File: core/src/main/kotlin/online/afeibaili/mchat/socket/Heartbeat.kt
package online.afeibaili.mchat.socket

import kotlinx.coroutines.*
import java.io.Closeable

/**
 * 心跳类
 *
 *@author AfeiBaili
 *@version 2025/11/3 18:46
 */

class Heartbeat(action: () -> Unit, catch: (exception: Throwable) -> Unit = {}) :
    Closeable {
    private val scope = CoroutineScope(Dispatchers.IO)

    init {
        scope.launch {
            runCatching {
                while (isActive) {
                    action()
                    delay(60000 * 3)
                }
            }.onFailure { exception ->
                catch(exception)
            }
        }
    }

    override fun close() {
        scope.cancel()
    }
}